apollo-router/src/services/subgraph_service.rs

```diff
@@ -597,104 +593,205 @@ async fn call_http(
-            graphql::Response::from_bytes(&cloned_service_name, body).map_err(|error| {
-                FetchError::SubrequestMalformedResponse {
-                    service: cloned_service_name.clone(),
-                    reason: error.to_string(),
-                }
+    let resp = http::Response::from_parts(parts, graphql);
+    Ok(SubgraphResponse::new_from_response(resp, context))
+}
+
+async fn perform_fetch(
+    mut client: Decompression<Client<HttpsConnector<HttpConnector>>>,
+    service_name: &str,
+    request: Request<Body>,
+) -> Result<(Parts, Result<Bytes, FetchError>), FetchError> {
+    let response = match client.call(request).await {
+        Err(err) => {
+            tracing::error!(fetch_error = format!("{err:?}").as_str());
+            return Err(FetchError::SubrequestHttpError {
+                status_code: None,
+                service: service_name.to_string(),
+                reason: err.to_string(),
+            });
+        }
+        Ok(response) => response,
+    };
+    // Keep our parts, we'll need them later
+    let (parts, body) = response.into_parts();
+
+    // Only try to get the body bytes if the content type was one of the accepted types
+    let body_bytes = match parts.headers.get(header::CONTENT_TYPE).map(|v| v.to_str()) {
+        Some(Ok(content_type))
+            if !content_type.contains(APPLICATION_JSON.essence_str())
+                && !content_type.contains(GRAPHQL_JSON_RESPONSE_HEADER_VALUE) =>
+        {
+            Err(FetchError::SubrequestHttpError {
+                status_code: Some(parts.status.as_u16()),
+                service: service_name.to_string(),
+                reason: format!("subgraph didn't return JSON (expected content-type: {} or content-type: {GRAPHQL_JSON_RESPONSE_HEADER_VALUE}; found content-type: {content_type:?})", APPLICATION_JSON.essence_str()),
             })
-        })?;
+        }
+        None | Some(Err(_)) => {
+            Err(FetchError::SubrequestHttpError {
+                status_code: Some(parts.status.as_u16()),
+                service: service_name.to_string(),
+                reason: format!("subgraph didn't return JSON (expected content-type: {} or content-type: {GRAPHQL_JSON_RESPONSE_HEADER_VALUE})", APPLICATION_JSON.essence_str()),
+            })
+        }
 
-    let resp = http::Response::from_parts(parts, graphql);
+        _ => {
+            match hyper::body::to_bytes(body)
+                .instrument(tracing::debug_span!("aggregate_response_data"))
+                .await
+            {
+                Err(err) => {
+                    tracing::error!(fetch_error = format!("{err:?}").as_str());
+                    Err(FetchError::SubrequestHttpError {
+                        status_code: None,
+                        service: service_name.to_string(),
+                        reason: err.to_string(),
+                    })
+                }
+                Ok(body) => Ok(body),
+            }
+        }
```

Comment:
Can't we use a `map_err` for this ?

Reply:
[7534cc6](https://github.com/apollographql/router/pull/3328/commits/7534cc6607c769c903bd62590d9d5a0c2c338b4f)